pkg/agent/controller/networkpolicy/fqdn_test.go

```diff
@@ -0,0 +1,286 @@
+		previouslyAddedRules    []fqdnRuleAddArgs
+		existingDNSCache        map[string]dnsMeta
+		ruleID                  string
+		fqdns                   []string
+		finalSelectorToRuleIDs  map[fqdnSelectorItem]sets.String
+		finalFQDNToSelectorItem map[string]map[fqdnSelectorItem]struct{}
+		addressRemoved          bool
+	}{
+		{
+			"test-1",
+			[]fqdnRuleAddArgs{
+				{
+					"mockRule1",
+					[]string{"test.antrea.io"},
+					sets.NewInt32(1),
+				},
+			},
+			map[string]dnsMeta{
+				"test.antrea.io": {},
+			},
+			"mockRule1",
+			[]string{"test.antrea.io"},
+			map[fqdnSelectorItem]sets.String{},
+			map[string]map[fqdnSelectorItem]struct{}{},
+			true,
+		},
+		{
+			"test-2",
+			[]fqdnRuleAddArgs{
+				{
+					"mockRule1",
+					[]string{"test.antrea.io"},
+					sets.NewInt32(1),
+				},
+				{
+					"mockRule2",
+					[]string{"test.antrea.io"},
+					sets.NewInt32(2),
+				},
+			},
+			map[string]dnsMeta{
+				"test.antrea.io": {},
+			},
+			"mockRule1",
+			[]string{"test.antrea.io"},
+			map[fqdnSelectorItem]sets.String{
+				selectorItem1: sets.NewString("mockRule2"),
+			},
+			map[string]map[fqdnSelectorItem]struct{}{
+				"test.antrea.io": {
+					selectorItem1: struct{}{},
+				},
+			},
+			true,
+		},
+		{
+			"test-3",
+			[]fqdnRuleAddArgs{
+				{
+					"mockRule1",
+					[]string{"test.antrea.io"},
+					sets.NewInt32(1),
+				},
+				{
+					"mockRule2",
+					[]string{"*antrea.io"},
+					sets.NewInt32(2),
+				},
+			},
+			map[string]dnsMeta{
+				"test.antrea.io": {},
+			},
+			"mockRule1",
+			[]string{"test.antrea.io"},
+			map[fqdnSelectorItem]sets.String{
+				selectorItem2: sets.NewString("mockRule2"),
+			},
+			map[string]map[fqdnSelectorItem]struct{}{
+				"test.antrea.io": {
+					selectorItem2: struct{}{},
+				},
+			},
+			true,
+		},
+		{
+			"test-4",
+			[]fqdnRuleAddArgs{
+				{
+					"mockRule1",
+					[]string{"maps.google.com"},
+					sets.NewInt32(1),
+				},
+				{
+					"mockRule2",
+					[]string{"*antrea.io$"},
+					sets.NewInt32(2),
+				},
+			},
+			map[string]dnsMeta{
+				"test.antrea.io":  {},
+				"maps.google.com": {},
+			},
+			"mockRule2",
+			[]string{"*antrea.io$"},
+			map[fqdnSelectorItem]sets.String{
+				selectorItem3: sets.NewString("mockRule1"),
+			},
+			map[string]map[fqdnSelectorItem]struct{}{
+				"maps.google.com": {
+					selectorItem3: struct{}{},
+				},
+			},
+			true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			controller := gomock.NewController(t)
+			defer controller.Finish()
+			f, c := newMockFQDNController(controller)
+			c.EXPECT().AddAddressToDNSConjunction(dnsInterceptRuleID, gomock.Any()).Times(len(tt.previouslyAddedRules))
+			f.dnsEntryCache = tt.existingDNSCache
+			if tt.addressRemoved {
+				c.EXPECT().DeleteAddressFromDNSConjunction(dnsInterceptRuleID, gomock.Any()).Times(1)
+			}
+			for _, r := range tt.previouslyAddedRules {
+				f.addFQDNRule(r.ruleID, r.fqdns, r.podOFAddresses)
```

Comment:
I think the returned error should be checked for easier troubleshooting in the future